### ISSUE TITLE
Support complex types on entity types participating in inheritance hierarchies (TPH/TPT/TPC)

### ISSUE THREAD

**Issue by dvdwouwe** [dotnet/efcore#35025]
### Include your code

The combination of inheritance mapping and the use of Complex types, gives strange behavior in TPC mapping.
I tried TPH and TPT (line 118), and there everything works as expected.

What I see is:
- model seems ok, but the creation of tables is not ok -> columns of the complex type are missing
- If I run polymorphic queries I get an exception in TPC, works fine in TPH and TPT

See https://github.com/dvdwouwe/danny-playground-ef/tree/main for reproducing the behavior we see.
Commit SHA: d194a6f9a7a783a5579a03ed9ceb764955e0dd2d

### Include provider and version information

EF Core version: 8.0.10
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 8.0.403
Operating system: Windows 11 
IDE: Rider

**Comment by cincuranet** [comment]
Not `Microsoft.EntityFrameworkCore.SqlServer`, repros on `Microsoft.EntityFrameworkCore.Sqlite` as well. Repros on `9.0.0-rc.2.24474.1`.

<details>
<summary>Smaller repro</summary>

```csharp
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Logging;

await using var context = new MyDbContext();
await context.Database.EnsureDeletedAsync();
await context.Database.EnsureCreatedAsync();
await context
    .Set<RealEvent>()
    .Where(e => e.Knowledge.To == null)
    .ToListAsync();

public abstract class EventBase
{
    protected EventBase(int id)
    {
        Id = id;
    }

    public int Id { get; set; }
    public Period Knowledge { get; set; }
}

public class RealEvent : EventBase
{
    public RealEvent(int id) : base(id)
    { }
}

public class Period
{
    public DateTimeOffset From { get; set; }
    public DateTimeOffset? To { get; set; }
}

public class MyDbContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            .EnableDetailedErrors()
            .EnableSensitiveDataLogging()
            .UseSqlite()
            .LogTo(Console.WriteLine, LogLevel.Information);

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        base.OnModelCreating(modelBuilder);

        modelBuilder.Entity<EventBase>(builder =>
        {
            builder.ComplexProperty(e => e.Knowledge);
            builder.UseTpcMappingStrategy();
        });
        modelBuilder.Entity<RealEvent>();
    }
}
```
</details>

The table created is only:

```sql
      CREATE TABLE "RealEvent" (
          "Id" INTEGER NOT NULL CONSTRAINT "PK_RealEvent" PRIMARY KEY
      );
```

Also, query throws exception:

```text
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.Single[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.SqlExpressions.SelectExpression.GenerateComplexPropertyShaperExpression(StructuralTypeProjectionExpression containerProjection, IComplexProperty complexProperty)
   at Microsoft.EntityFrameworkCore.Query.StructuralTypeProjectionExpression.BindComplexProperty(IComplexProperty complexProperty)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.BindComplexProperty(StructuralTypeReferenceExpression typeReference, IComplexProperty complexProperty)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.TryBindMember(Expression source, MemberIdentity member, Expression& expression, IPropertyBase& property)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.TryBindMember(Expression source, MemberIdentity member, Expression& expression)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.VisitMember(MemberExpression memberExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.VisitMember(MemberExpression memberExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.VisitBinary(BinaryExpression binaryExpression)
   at Microsoft.EntityFrameworkCore.Sqlite.Query.Internal.SqliteSqlTranslatingExpressionVisitor.VisitBinary(BinaryExpression binaryExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.TranslateInternal(Expression expression, Boolean applyDefaultTypeMapping)
   at Microsoft.EntityFrameworkCore.Query.RelationalSqlTranslatingExpressionVisitor.Translate(Expression expression, Boolean applyDefaultTypeMapping)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateExpression(Expression expression, Boolean applyDefaultTypeMapping)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateLambdaExpression(ShapedQueryExpression shapedQueryExpression, LambdaExpression lambdaExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.TranslateWhere(ShapedQueryExpression source, LambdaExpression predicate)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.RelationalQueryableMethodTranslatingExpressionVisitor.VisitMethodCall(MethodCallExpression methodCallExpression)
   at Microsoft.EntityFrameworkCore.Query.QueryableMethodTranslatingExpressionVisitor.Translate(Expression expression)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutorExpression[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.QueryCompilationContext.CreateQueryExecutor[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Storage.Database.CompileQuery[TResult](Expression query, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.CompileQueryCore[TResult](IDatabase database, Expression query, IModel model, Boolean async)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass11_0`1.<ExecuteCore>b__0()
   at Microsoft.EntityFrameworkCore.Query.Internal.CompiledQueryCache.GetOrAddQuery[TResult](Object cacheKey, Func`1 compiler)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteCore[TResult](Expression query, Boolean async, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.ExecuteAsync[TResult](Expression query, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.ExecuteAsync[TResult](Expression expression, CancellationToken cancellationToken)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryable`1.GetAsyncEnumerator(CancellationToken cancellationToken)
   at System.Runtime.CompilerServices.ConfiguredCancelableAsyncEnumerable`1.GetAsyncEnumerator()
   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.ToListAsync[TSource](IQueryable`1 source, CancellationToken cancellationToken)
```

**Comment by maumar** [comment]
When trying to bind complex property we are looking for table mapping so that we know which table/columns to bind. However the table mapping is missing here - We build those as part of RelationalModel.Create. Here, the complex type is defined on the abstract base type, and we skip those when creating table mappings. When processing the derived type we only look at declared complex types and so we miss the type defined on the base. As a result the complex property ends up with no table mapping

**Comment by dvdwouwe** [comment]
Hi,

Actually this is blocking us. We have a rather huge (potential) project by a large company, but we need an example that is production ready. The expectations for this in August 2025. If it is successful, it will be one of our biggest projects ever.

TPH mapping is not very suitable for this:

* too many null columns in this case
* complex unique constraints, always take into account the discriminator field 
* too complex indices
* if db admins see such a huge table, they really don't like it

TPT is too slow, because of the huge number of abstract subclasses.

**Comment by AndriySvyryd** [comment]
@dvdwouwe There are two workarounds that you can consider using for now:

1. Use owned types instead of complex types for this case. This means that they will get a shadow PK, so you might need to do extra work when you are attaching entities containing them
2. Ignore the complex properties on the base type and only configure it on the leaf types. This would mean that you can't use these properties in queries at the base type level:
```C#
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.Entity<EventBase>(builder =>
    {
        builder.Ignore(e => e.Knowledge);
        builder.UseTpcMappingStrategy();
    });
    modelBuilder.Entity<EventWithName>(builder =>
    {
        builder.ComplexProperty(e => e.Knowledge).IsRequired();
    });
    modelBuilder.Entity<EventWithPartnerType1>(builder =>
    {
        builder.ComplexProperty(e => e.Knowledge).IsRequired();
    });
    modelBuilder.Entity<EventWithPartnerType2>(builder =>
    {
        builder.ComplexProperty(e => e.Knowledge).IsRequired();
    });
}
```

**Comment by dvdwouwe** [comment]
@AndriySvyryd ,

Thanks for the suggestions.

I have already experimented with both approaches:

1. **Owned types**: It seems they are only available on leaf nodes, even if they are defined on a non-leaf type.
2. **Complex properties**: These are also limited to leaf nodes.

Both workarounds present the same significant issue:

- I cannot use **polymorphic queries**. I have many subclasses and need to query a tree of classes. In both cases, the workaround doesn't solve this problem.

The only solution (workaround) that works for me is using **TPH (Table Per Hierarchy)**. With TPH, the complex properties are recognized correctly, and I can use polymorphic queries.

**Comment by dvdwouwe** [comment]
Is there more information about this issue? Will it be fixed in the next major version 10?

**Comment by AndriySvyryd** [comment]
> Will it be fixed in the next major version 10?

It's in the 11.0.0 Milestone

**Comment by roji** [comment]
@AndriySvyryd do you think there's anything specific more we need to do here after #37410? It feels like we're probably a bit light on testing (both query and non-query) - at least some smoke testing/bashing would probably be advisable. But otherwise do you think we should close this?

**Comment by AndriySvyryd** [comment]
Yes, I think all the functionality is now in-place, this will naturally get more tests as we work on https://github.com/dotnet/efcore/issues/36646, https://github.com/dotnet/efcore/issues/28443 and https://github.com/dotnet/efcore/issues/34627

**Comment by roji** [comment]
Thanks. I do think we should dedicate some specific additional testing/bashing to complex JSON on hierarchies for 11.

Looking at #28443 (which is a highly-voted issue), I think although up to now it tracked owned entities, people were just asking for a general way to have JSON properties on TPT/TPC entity types - we just provided that via complex types. With the direction of possibly obsoleting owned JSON altogether (#37290), I don't see us working on it anytime soon, and the highly-voted issue is currently a bit confusing and gives the impression we don't support any JSON on TPC/TPT.

So I'll go ahead and convert that issue to track what we've done (and close it), and close as a duplicate of that; I'll open a separate issue to track owned JSON support on TPT/TPC entities, so we can gather feedback on whether anybody cares about that specifically.
